actix-web-codegen/src/lib.rs

```diff
@@ -1,118 +1,171 @@
 #![recursion_limit = "512"]
+//! Actix-web codegen module
+//!
+//! Generators for routes and scopes
+//!
+//! ## Route
+//!
+//! Macros:
+//!
+//! - [handler](attr.handler.html)
+//! - [get](attr.get.html)
+//! - [post](attr.post.html)
+//! - [put](attr.put.html)
+//! - [delete](attr.delete.html)
+//!
+//! ### Attributes:
+//!
+//! - `"path"` - Raw literal string with path for which to register handle. Mandatory.
+//! - `async` - Attribute to indicate that registered function is asynchronous.
+//! - `sync` - Attribute to indicate that registered as regular function.
+//! - `guard="function_name"` - Registers function as guard using `actix_web::guard::fn_guard`
+//!
+//! ## Scope
+//!
+//! Macros:
+//!
+//! - [scope](attr.scope.html)
+//!
+//! ### Attributes:
+//!
+//! - `"path"` - Raw literal string with path for which to register handle. Mandatory.
+//! - `default="function_name"` - Registers function default resource factory. Gets passed to `Scope::default_resource`.
+//! - `guard="function_name"` - Registers function as guard using `actix_web::guard::fn_guard`
+//!
+//! ## Notes
+//!
+//! Function name can be specified as any expression that is going to be accessible to the generate
+//! code (e.g `my_guard` or `my_module::my_guard`)
+//!
+//! ## Example:
+//!
+//! ```rust
+//! use actix_web::HttpResponse;
+//! use actix_web_codegen::get;
+//! use futures::{future, Future};
+//!
+//! #[get("/test", async)]
+//! fn async_test() -> impl Future<Item=HttpResponse, Error=actix_web::Error> {
+//!     future::ok(HttpResponse::Ok().finish())
+//! }
+//! ```
 
 extern crate proc_macro;
 
+mod route;
+mod scope;
+
 use proc_macro::TokenStream;
-use quote::quote;
 use syn::parse_macro_input;
 
-/// #[get("path")] attribute
+/// Creates route handler without attached method guard.
+///
+/// Syntax: `#[handler("path"[, attributes])]`
+///
+/// ## Attributes:
+///
+/// - `"path"` - Raw literal string with path for which to register handler. Mandatory.
+/// - `async` - Attribute to indicate that registered function is asynchronous.
+/// - `sync` - Attribute to indicate that registered as regular function.
```

Comment:
cant you auto detect async/sync functions by `impl Future`?

Reply:
I do auto detect it, if you take a look at tests, some of them don't mark as async/sync
These markers are mostly for cases when user want to override it

Reply:
If you auto detect then remove attributes.

Reply:
You cannot auto-detect everything, user might have `MyBoxedAlias` type and there is no way to reliable detect whether it is future or not

Reply:
I'll remove `sync` since it is default

Reply:
we dont need to support 100% of use cases,  80% is more than enough, but those 80% should be easy to use.

Reply:
Remove `sync` but `async` remains. Though it is optional for use